File: bubus-ts/src/bridges.ts
import { BaseEvent } from './base_event.js'
import { EventBus } from './event_bus.js'
import type { EventClass, EventHandlerFunction, EventKey, UntypedEventHandlerFunction } from './types.js'

type EndpointScheme = 'unix' | 'http' | 'https'

type ParsedEndpoint = {
  raw: string
  scheme: EndpointScheme
  host?: string
  port?: number
  path?: string
}

export type HTTPEventBridgeOptions = {
  send_to?: string | null
  listen_on?: string | null
  name?: string
}

const isNodeRuntime = (): boolean => {
  const maybe_process = (globalThis as { process?: { versions?: { node?: string } } }).process
  return typeof maybe_process?.versions?.node === 'string'
}

const isBrowserRuntime = (): boolean => !isNodeRuntime() && typeof globalThis.window !== 'undefined'

const randomSuffix = (): string => Math.random().toString(36).slice(2, 10)
const UNIX_SOCKET_MAX_PATH_CHARS = 90

const parseEndpoint = (raw_endpoint: string): ParsedEndpoint => {
  let parsed: URL
  try {
    parsed = new URL(raw_endpoint)
  } catch {
    throw new Error(`Invalid endpoint URL: ${raw_endpoint}`)
  }

  const protocol = parsed.protocol.replace(/:$/, '').toLowerCase()
  if (protocol !== 'unix' && protocol !== 'http' && protocol !== 'https') {
    throw new Error(`Unsupported endpoint scheme: ${raw_endpoint}`)
  }

  if (protocol === 'unix') {
    const socket_path = decodeURIComponent(parsed.pathname || '')
    if (!socket_path) {
      throw new Error(`Invalid unix endpoint (missing socket path): ${raw_endpoint}`)
    }
    const socket_path_len = new TextEncoder().encode(socket_path).length
    if (socket_path_len > UNIX_SOCKET_MAX_PATH_CHARS) {
      throw new Error(`Unix socket path is too long (${socket_path_len} chars), max is ${UNIX_SOCKET_MAX_PATH_CHARS}: ${socket_path}`)
    }
    return { raw: raw_endpoint, scheme: 'unix', path: socket_path }
  }

  if (!parsed.hostname) {
    throw new Error(`Invalid HTTP endpoint (missing hostname): ${raw_endpoint}`)
  }

  const default_port = protocol === 'https' ? 443 : 80
  return {
    raw: raw_endpoint,
    scheme: protocol,
    host: parsed.hostname,
    port: parsed.port ? Number(parsed.port) : default_port,
    path: `${parsed.pathname || '/'}${parsed.search || ''}`,
  }
}

const importNodeModule = async (specifier: string): Promise<any> => {
  const dynamic_import = Function('module_name', 'return import(module_name)') as (module_name: string) => Promise<unknown>
  return dynamic_import(specifier) as Promise<any>
}

class _EventBridge {
  readonly send_to: ParsedEndpoint | null
  readonly listen_on: ParsedEndpoint | null
  readonly name: string

  protected readonly inbound_bus: EventBus
  private start_promise: Promise<void> | null
  private node_server: any | null

  constructor(send_to?: string | null, listen_on?: string | null, name?: string) {
    this.send_to = send_to ? parseEndpoint(send_to) : null
    this.listen_on = listen_on ? parseEndpoint(listen_on) : null
    this.name = name ?? `EventBridge_${randomSuffix()}`
    this.inbound_bus = new EventBus(this.name)
    this.start_promise = null
    this.node_server = null

    if (this.listen_on && isBrowserRuntime()) {
      throw new Error(`${this.constructor.name} listen_on is not supported in browser runtimes`)
    }

    this.dispatch = this.dispatch.bind(this)
    this.emit = this.emit.bind(this)
    this.on = this.on.bind(this)
  }

  on<T extends BaseEvent>(event_key: EventClass<T>, handler: EventHandlerFunction<T>): void
  on<T extends BaseEvent>(event_key: string | '*', handler: UntypedEventHandlerFunction<T>): void
  on(event_key: EventKey | '*', handler: EventHandlerFunction | UntypedEventHandlerFunction): void {
    this.ensureListenerStarted()
    if (typeof event_key === 'string') {
      this.inbound_bus.on(event_key, handler as UntypedEventHandlerFunction<BaseEvent>)
      return
    }
    this.inbound_bus.on(event_key as EventClass<BaseEvent>, handler as EventHandlerFunction<BaseEvent>)
  }

  async dispatch<T extends BaseEvent>(event: T): Promise<void> {
    if (!this.send_to) {
      throw new Error(`${this.constructor.name}.dispatch() requires send_to`)
    }

    const payload = event.toJSON()

    if (this.send_to.scheme === 'unix') {
      await this.sendUnix(this.send_to, payload)
      return
    }

    await this.sendHttp(this.send_to, payload)
  }

  async emit<T extends BaseEvent>(event: T): Promise<void> {
    return this.dispatch(event)
  }

  async start(): Promise<void> {
    if (!this.listen_on) return
    if (this.node_server) return
    if (this.start_promise) {
      await this.start_promise
      return
    }

    if (!isNodeRuntime()) {
      throw new Error(`${this.constructor.name} listen_on is only supported in Node.js runtimes`)
    }

    const launch = (async () => {
      const endpoint = this.listen_on
      if (!endpoint) return

      if (endpoint.scheme === 'unix') {
        await this.startUnixListener(endpoint)
        return
      }

      if (endpoint.scheme !== 'http') {
        throw new Error(`listen_on only supports unix:// or http:// endpoints, got: ${endpoint.raw}`)
      }

      await this.startHttpListener(endpoint)
    })()
    this.start_promise = launch

    try {
      await launch
    } finally {
      if (this.start_promise === launch) {
        this.start_promise = null
      }
    }
  }

  async close(): Promise<void> {
    if (this.start_promise) {
      await Promise.allSettled([this.start_promise])
      this.start_promise = null
    }

    if (this.node_server) {
      const server = this.node_server
      await new Promise<void>((resolve) => {
        server.close(() => resolve())
      })
      this.node_server = null
    }

    this.inbound_bus.destroy()
  }

  private ensureListenerStarted(): void {
    if (!this.listen_on || this.node_server || this.start_promise) {
      return
    }
    void this.start().catch((error: unknown) => {
      console.error('[bubus] EventBridge failed to start listener', error)
    })
  }

  private async handleIncomingPayload(payload: unknown): Promise<void> {
    const parsed_event = BaseEvent.fromJSON(payload)
    const existing_event = EventBus._all_instances.findEventById(parsed_event.event_id)
    const event = existing_event ?? parsed_event.reset()
    this.inbound_bus.dispatch(event)
  }

  private async sendHttp(endpoint: ParsedEndpoint, payload: unknown): Promise<void> {
    const response = await fetch(endpoint.raw, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(payload),
    })
    if (!response.ok) {
      throw new Error(`IPC HTTP send failed with status ${response.status}: ${endpoint.raw}`)
    }
  }

  private async sendUnix(endpoint: ParsedEndpoint, payload: unknown): Promise<void> {
    if (!isNodeRuntime()) {
      throw new Error('unix:// send_to is only supported in Node.js runtimes')
    }

    const socket_path = endpoint.path
    if (!socket_path) {
      throw new Error(`Invalid unix endpoint: ${endpoint.raw}`)
    }

    const node_net = await importNodeModule('node:net')
    await new Promise<void>((resolve, reject) => {
      const socket = node_net.createConnection(socket_path, () => {
        socket.end(`${JSON.stringify(payload)}\n`)
      })
      socket.on('error', (error: unknown) => reject(error))
      socket.on('close', () => resolve())
    })
  }

  private async startHttpListener(endpoint: ParsedEndpoint): Promise<void> {
    const node_http = await importNodeModule('node:http')
    const expected_path = endpoint.path || '/'

    this.node_server = node_http.createServer((req: any, res: any) => {
      const method = (req.method || '').toUpperCase()
      const request_url = String(req.url || '/')

      if (method !== 'POST') {
        res.statusCode = 405
        res.end('method not allowed')
        return
      }
      if (request_url !== expected_path) {
        res.statusCode = 404
        res.end('not found')
        return
      }

      let body = ''
      req.setEncoding('utf8')
      req.on('data', (chunk: string) => {
        body += chunk
      })
      req.on('end', () => {
        let parsed_payload: unknown
        try {
          parsed_payload = JSON.parse(body)
        } catch {
          res.statusCode = 400
          res.end('invalid json')
          return
        }

        void this.handleIncomingPayload(parsed_payload)
          .then(() => {
            res.statusCode = 202
            res.end('accepted')
          })
          .catch((error: unknown) => {
            res.statusCode = 500
            res.end('failed to process event')
            console.error('[bubus] EventBridge HTTP listener error', error)
          })
      })
    })

    await new Promise<void>((resolve, reject) => {
      this.node_server.once('error', (error: unknown) => reject(error))
      this.node_server.listen(endpoint.port, endpoint.host, () => resolve())
    })
  }

  private async startUnixListener(endpoint: ParsedEndpoint): Promise<void> {
    const socket_path = endpoint.path
    if (!socket_path) {
      throw new Error(`Invalid unix endpoint: ${endpoint.raw}`)
    }

    const node_net = await importNodeModule('node:net')
    const node_fs = await importNodeModule('node:fs')

    try {
      await node_fs.promises.unlink(socket_path)
    } catch (error: unknown) {
      const code = (error as { code?: string }).code
      if (code !== 'ENOENT') {
        throw error
      }
    }

    this.node_server = node_net.createServer((socket: any) => {
      let buffer = ''
      socket.setEncoding('utf8')
      socket.on('data', (chunk: string) => {
        buffer += chunk
        while (true) {
          const newline_index = buffer.indexOf('\n')
          if (newline_index < 0) break
          const line = buffer.slice(0, newline_index).trim()
          buffer = buffer.slice(newline_index + 1)
          if (!line) continue
          try {
            const parsed_payload = JSON.parse(line)
            void this.handleIncomingPayload(parsed_payload)
          } catch {
            // Ignore malformed lines and continue reading next frames.
          }
        }
      })
      socket.on('end', () => {
        const remainder = buffer.trim()
        if (!remainder) return
        try {
          const parsed_payload = JSON.parse(remainder)
          void this.handleIncomingPayload(parsed_payload)
        } catch {
          // Ignore malformed trailing frame.
        }
      })
    })

    await new Promise<void>((resolve, reject) => {
      this.node_server.once('error', (error: unknown) => reject(error))
      this.node_server.listen(socket_path, () => resolve())
    })
  }
}

export class HTTPEventBridge extends _EventBridge {
  constructor(send_to?: string | null, listen_on?: string | null, name?: string)
  constructor(options?: HTTPEventBridgeOptions)
  constructor(send_to_or_options?: string | null | HTTPEventBridgeOptions, listen_on?: string | null, name?: string) {
    const options: HTTPEventBridgeOptions =
      typeof send_to_or_options === 'object'
        ? (send_to_or_options ?? {})
        : { send_to: send_to_or_options ?? undefined, listen_on: listen_on ?? undefined, name }

    if (options.send_to && parseEndpoint(options.send_to).scheme === 'unix') {
      throw new Error('HTTPEventBridge send_to must be http:// or https://')
    }
    if (options.listen_on && parseEndpoint(options.listen_on).scheme !== 'http') {
      throw new Error('HTTPEventBridge listen_on must be http://')
    }

    super(options.send_to, options.listen_on, options.name ?? `HTTPEventBridge_${randomSuffix()}`)
  }
}

export class SocketEventBridge extends _EventBridge {
  constructor(path?: string | null, name?: string) {
    const normalized = path ? (path.startsWith('unix://') ? path.slice(7) : path) : null
    if (normalized === '') {
      throw new Error('SocketEventBridge path must not be empty')
    }

    const endpoint = normalized ? `unix://${normalized}` : null
    super(endpoint, endpoint, name ?? `SocketEventBridge_${randomSuffix()}`)
  }
}

export { NATSEventBridge } from './bridge_nats.js'
export { RedisEventBridge } from './bridge_redis.js'
export { PostgresEventBridge } from './bridge_postgres.js'
export { JSONLEventBridge } from './bridge_jsonl.js'
export { SQLiteEventBridge } from './bridge_sqlite.js'

File: bubus/bridges.py
"""IPC bridges for forwarding EventBus instances over HTTP or unix sockets."""

from __future__ import annotations

import asyncio
import importlib
import json
import logging
from collections.abc import Callable
from pathlib import Path
from typing import TYPE_CHECKING, Any, Literal
from urllib.parse import urlparse
from urllib.request import Request, urlopen

from anyio import Path as AnyPath
from uuid_extensions import uuid7str

from bubus.event_bus import EventBus, EventPatternType, in_handler_context
from bubus.models import BaseEvent

logger = logging.getLogger('bubus.bridges')
UNIX_SOCKET_MAX_PATH_CHARS = 90
__all__ = [
    'HTTPEventBridge',
    'SocketEventBridge',
    'NATSEventBridge',
    'RedisEventBridge',
    'PostgresEventBridge',
    'JSONLEventBridge',
    'SQLiteEventBridge',
]

if TYPE_CHECKING:
    from .bridge_jsonl import JSONLEventBridge
    from .bridge_nats import NATSEventBridge
    from .bridge_postgres import PostgresEventBridge
    from .bridge_redis import RedisEventBridge
    from .bridge_sqlite import SQLiteEventBridge

_LAZY_BRIDGE_MODULES: dict[str, str] = {
    'NATSEventBridge': '.bridge_nats',
    'RedisEventBridge': '.bridge_redis',
    'PostgresEventBridge': '.bridge_postgres',
    'JSONLEventBridge': '.bridge_jsonl',
    'SQLiteEventBridge': '.bridge_sqlite',
}


class _Endpoint:
    def __init__(
        self,
        raw: str,
        scheme: Literal['unix', 'http', 'https'],
        *,
        host: str | None = None,
        port: int | None = None,
        path: str | None = None,
    ):
        self.raw = raw
        self.scheme = scheme
        self.host = host
        self.port = port
        self.path = path


def _parse_endpoint(raw_endpoint: str) -> _Endpoint:
    parsed = urlparse(raw_endpoint)
    scheme = parsed.scheme.lower()

    if scheme == 'unix':
        socket_path = parsed.path or parsed.netloc
        if not socket_path:
            raise ValueError(f'Invalid unix endpoint (missing socket path): {raw_endpoint}')
        socket_path_len = len(socket_path.encode('utf-8'))
        if socket_path_len > UNIX_SOCKET_MAX_PATH_CHARS:
            raise ValueError(
                f'Unix socket path is too long ({socket_path_len} chars), max is {UNIX_SOCKET_MAX_PATH_CHARS}: {socket_path}'
            )
        return _Endpoint(raw_endpoint, 'unix', path=socket_path)

    if scheme == 'http':
        if not parsed.hostname:
            raise ValueError(f'Invalid HTTP endpoint (missing hostname): {raw_endpoint}')
        request_path = parsed.path or '/'
        if parsed.query:
            request_path = f'{request_path}?{parsed.query}'
        port = parsed.port if parsed.port is not None else 80
        return _Endpoint(raw_endpoint, 'http', host=parsed.hostname, port=port, path=request_path)

    if scheme == 'https':
        if not parsed.hostname:
            raise ValueError(f'Invalid HTTP endpoint (missing hostname): {raw_endpoint}')
        request_path = parsed.path or '/'
        if parsed.query:
            request_path = f'{request_path}?{parsed.query}'
        port = parsed.port if parsed.port is not None else 443
        return _Endpoint(raw_endpoint, 'https', host=parsed.hostname, port=port, path=request_path)

    raise ValueError(f'Unsupported endpoint scheme: {raw_endpoint}')


class EventBridge:
    """Shared bridge implementation exposing EventBus-like on/emit/dispatch."""

    def __init__(
        self,
        send_to: str | None = None,
        listen_on: str | None = None,
        *,
        name: str | None = None,
    ):
        self.send_to = _parse_endpoint(send_to) if send_to else None
        self.listen_on = _parse_endpoint(listen_on) if listen_on else None
        internal_name = name or f'EventBridge_{uuid7str()[-8:]}'
        self._inbound_bus = EventBus(name=internal_name, max_history_size=0)

        self._server: asyncio.AbstractServer | None = None
        self._start_lock = asyncio.Lock()
        self._listen_socket_path: Path | None = None
        self._autostart_task: asyncio.Task[None] | None = None

    def on(self, event_pattern: EventPatternType, handler: Callable[[BaseEvent[Any]], Any]) -> None:
        self._ensure_listener_started()
        self._inbound_bus.on(event_pattern, handler)

    async def dispatch(self, event: BaseEvent[Any]) -> BaseEvent[Any] | None:
        if self.send_to is None:
            raise RuntimeError(f'{self.__class__.__name__}.dispatch() requires send_to=...')

        payload = event.model_dump(mode='json')

        if self.send_to.scheme == 'unix':
            await self._send_unix(self.send_to, payload)
        else:
            await self._send_http(self.send_to, payload)

        if in_handler_context():
            return None
        return event

    async def emit(self, event: BaseEvent[Any]) -> BaseEvent[Any] | None:
        return await self.dispatch(event)

    async def start(self) -> None:
        if self.listen_on is None or self._server is not None:
            return

        async with self._start_lock:
            if self._server is not None:
                return

            endpoint = self.listen_on
            assert endpoint is not None
            if endpoint.scheme == 'unix':
                socket_path = Path(endpoint.path or '')
                if not socket_path.is_absolute():
                    raise ValueError(f'unix listen_on path must be absolute, got: {endpoint.raw}')
                socket_path.parent.mkdir(parents=True, exist_ok=True)
                async_socket_path = AnyPath(socket_path)
                if await async_socket_path.exists():
                    await async_socket_path.unlink()
                self._listen_socket_path = socket_path
                self._server = await asyncio.start_unix_server(self._handle_unix_client, path=str(socket_path))
                return

            if endpoint.scheme != 'http':
                raise ValueError(f'listen_on only supports unix:// or http:// endpoints, got: {endpoint.raw}')
            assert endpoint.host is not None
            assert endpoint.port is not None
            self._server = await asyncio.start_server(self._handle_http_client, host=endpoint.host, port=endpoint.port)

    async def close(self, *, clear: bool = True) -> None:
        if self._autostart_task is not None:
            await asyncio.gather(self._autostart_task, return_exceptions=True)
            self._autostart_task = None

        if self._server is not None:
            self._server.close()
            await self._server.wait_closed()
            self._server = None

        if self._listen_socket_path and self._listen_socket_path.exists():
            self._listen_socket_path.unlink()
            self._listen_socket_path = None

        await self._inbound_bus.stop(clear=clear)

    def _ensure_listener_started(self) -> None:
        if self.listen_on is None or self._server is not None:
            return
        if self._autostart_task is not None and not self._autostart_task.done():
            return
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            return
        self._autostart_task = loop.create_task(self.start())

    async def _handle_unix_client(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
        try:
            while True:
                line = await reader.readline()
                if not line:
                    break
                if not line.strip():
                    continue
                await self._handle_incoming_bytes(line)
        finally:
            writer.close()
            await writer.wait_closed()

    async def _handle_http_client(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
        try:
            raw_headers = await reader.readuntil(b'\r\n\r\n')
        except asyncio.IncompleteReadError:
            await self._write_http_response(writer, status=400, body='incomplete request')
            return
        except asyncio.LimitOverrunError:
            await self._write_http_response(writer, status=400, body='headers too large')
            return

        header_lines = raw_headers.decode('utf-8', errors='replace').split('\r\n')
        if not header_lines or not header_lines[0]:
            await self._write_http_response(writer, status=400, body='missing request line')
            return

        parts = header_lines[0].split(' ', 2)
        if len(parts) != 3:
            await self._write_http_response(writer, status=400, body='invalid request line')
            return
        method, request_target, _version = parts

        headers: dict[str, str] = {}
        for line in header_lines[1:]:
            if not line:
                continue
            name, separator, value = line.partition(':')
            if not separator:
                continue
            headers[name.strip().lower()] = value.strip()

        if method.upper() != 'POST':
            await self._write_http_response(writer, status=405, body='method not allowed')
            return

        expected_path = (self.listen_on.path if self.listen_on else None) or '/'
        if request_target != expected_path:
            await self._write_http_response(writer, status=404, body='not found')
            return

        content_length = headers.get('content-length', '0')
        try:
            body_size = int(content_length)
        except ValueError:
            await self._write_http_response(writer, status=400, body='invalid content-length')
            return

        if body_size < 0:
            await self._write_http_response(writer, status=400, body='invalid content-length')
            return

        try:
            body = await reader.readexactly(body_size)
        except asyncio.IncompleteReadError:
            await self._write_http_response(writer, status=400, body='incomplete body')
            return

        try:
            await self._handle_incoming_bytes(body)
        except Exception as exc:  # pragma: no cover
            logger.exception('Failed to process inbound IPC event: %s', exc)
            await self._write_http_response(writer, status=500, body='failed to process event')
            return

        await self._write_http_response(writer, status=202, body='accepted')

    async def _handle_incoming_bytes(self, payload: bytes) -> None:
        message = json.loads(payload.decode('utf-8'))
        event = BaseEvent[Any].model_validate(message).event_reset()
        self._inbound_bus.dispatch(event)

    async def _send_unix(self, endpoint: _Endpoint, payload: dict[str, Any]) -> None:
        socket_path = endpoint.path or ''
        if not socket_path:
            raise ValueError(f'Invalid unix endpoint: {endpoint.raw}')

        _reader, writer = await asyncio.open_unix_connection(path=socket_path)
        writer.write(json.dumps(payload, separators=(',', ':')).encode('utf-8'))
        writer.write(b'\n')
        await writer.drain()
        writer.close()
        await writer.wait_closed()

    async def _send_http(self, endpoint: _Endpoint, payload: dict[str, Any]) -> None:
        payload_bytes = json.dumps(payload, separators=(',', ':')).encode('utf-8')
        request = Request(
            endpoint.raw,
            data=payload_bytes,
            headers={
                'content-type': 'application/json',
                'content-length': str(len(payload_bytes)),
            },
            method='POST',
        )

        def _post() -> int:
            with urlopen(request, timeout=10) as response:
                return int(response.status)

        status_code = await asyncio.to_thread(_post)
        if status_code < 200 or status_code >= 300:
            raise RuntimeError(f'IPC HTTP send failed with status {status_code}: {endpoint.raw}')

    @staticmethod
    async def _write_http_response(writer: asyncio.StreamWriter, *, status: int, body: str) -> None:
        reasons = {
            202: 'Accepted',
            400: 'Bad Request',
            404: 'Not Found',
            405: 'Method Not Allowed',
            500: 'Internal Server Error',
        }
        reason = reasons.get(status, 'OK')
        body_bytes = body.encode('utf-8')
        headers = [
            f'HTTP/1.1 {status} {reason}',
            f'content-length: {len(body_bytes)}',
            'content-type: text/plain; charset=utf-8',
            'connection: close',
            '',
            '',
        ]
        writer.write('\r\n'.join(headers).encode('utf-8'))
        writer.write(body_bytes)
        await writer.drain()
        writer.close()
        await writer.wait_closed()


class HTTPEventBridge(EventBridge):
    """Bridge events over HTTP(S) endpoints."""

    def __init__(self, send_to: str | None = None, listen_on: str | None = None, *, name: str | None = None):
        if send_to and _parse_endpoint(send_to).scheme == 'unix':
            raise ValueError('HTTPEventBridge send_to must be http:// or https://')
        if listen_on and _parse_endpoint(listen_on).scheme != 'http':
            raise ValueError('HTTPEventBridge listen_on must be http://')
        super().__init__(send_to=send_to, listen_on=listen_on, name=name or f'HTTPEventBridge_{uuid7str()[-8:]}')


class SocketEventBridge(EventBridge):
    """Bridge events over a unix domain socket path."""

    def __init__(self, path: str | None = None, *, name: str | None = None):
        if path is None:
            send_to = None
            listen_on = None
        else:
            normalized = path[7:] if path.startswith('unix://') else path
            if not normalized:
                raise ValueError('SocketEventBridge path must not be empty')
            send_to = f'unix://{normalized}'
            listen_on = f'unix://{normalized}'

        super().__init__(send_to=send_to, listen_on=listen_on, name=name or f'SocketEventBridge_{uuid7str()[-8:]}')


def __getattr__(name: str) -> Any:
    module_name = _LAZY_BRIDGE_MODULES.get(name)
    if module_name is None:
        raise AttributeError(name)
    module = importlib.import_module(module_name, __package__)
    value = getattr(module, name)
    globals()[name] = value
    return value
